internal/server/receiver_handlers.go

```diff
@@ -61,45 +76,48 @@ func (s *ReceiverServer) handlePayload() func(w http.ResponseWriter, r *http.Req
 		s.logger.Info(fmt.Sprintf("handling request: %s", digest))
 
 		var allReceivers apiv1.ReceiverList
-		err := s.kubeClient.List(ctx, &allReceivers)
+		err := s.kubeClient.List(ctx, &allReceivers, client.MatchingFields{
+			WebhookPathIndexKey: r.RequestURI,
+		}, client.Limit(1))
 		if err != nil {
 			s.logger.Error(err, "unable to list receivers")
 			w.WriteHeader(http.StatusBadRequest)
 			return
 		}
 
-		receivers := make([]apiv1.Receiver, 0)
-		for _, receiver := range allReceivers.Items {
-			if !receiver.Spec.Suspend &&
-				conditions.IsReady(&receiver) &&
-				receiver.Status.WebhookPath == fmt.Sprintf("%s%s", apiv1.ReceiverWebhookPath, digest) {
-				receivers = append(receivers, receiver)
-			}
-		}
-
-		if len(receivers) == 0 {
+		if len(allReceivers.Items) == 0 {
 			w.WriteHeader(http.StatusNotFound)
 			return
 		}
 
-		withErrors := false
-		for _, receiver := range receivers {
-			logger := s.logger.WithValues(
-				"reconciler kind", apiv1.ReceiverKind,
-				"name", receiver.Name,
-				"namespace", receiver.Namespace)
+		receiver := allReceivers.Items[0]
+		logger := s.logger.WithValues(
+			"reconciler kind", apiv1.ReceiverKind,
+			"name", receiver.Name,
+			"namespace", receiver.Namespace)
 
-			if err := s.validate(ctx, receiver, r); err != nil {
-				logger.Error(err, "unable to validate payload")
-				withErrors = true
-				continue
+		if receiver.Spec.Suspend || !conditions.IsReady(&receiver) {
+			err := errors.New("unable to process request")
+			if receiver.Spec.Suspend {
+				logger.Error(err, "receiver is suspended")
+			} else {
+				logger.Error(err, "receiver is not ready")
 			}
+			w.WriteHeader(http.StatusServiceUnavailable)
+			return
+		}
 
-			for _, resource := range receiver.Spec.Resources {
-				if err := s.requestReconciliation(ctx, logger, resource, receiver.Namespace); err != nil {
-					logger.Error(err, "unable to process resource")
-					withErrors = true
-				}
+		if err := s.validate(ctx, receiver, r); err != nil {
+			logger.Error(err, "unable to validate payload")
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		var withErrors bool
+		for _, resource := range receiver.Spec.Resources {
+			if err := s.requestReconciliation(ctx, logger, resource, receiver.Namespace); err != nil {
+				logger.Error(err, "unable to request reconciliation")
+				withErrors = true
 			}
 		}
 
```

Comment:
One nit: I think `withErrors` below should actually return a HTTP 500. As the request can be made again without making a modification to the request.

Which isn't true for HTTP 400: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400